src/libraries/Hooks.sol

```diff
@@ -57,50 +72,68 @@ library Hooks {
     function isValidHookAddress(IHooks hook, uint24 fee) internal pure returns (bool) {
+        // The hook can only have a flag to return a hook delta if it has the corresponding hook flag
+        if (
+            (!hook.hasPermission(BEFORE_SWAP_FLAG) && hook.hasPermission(BEFORE_SWAP_RETURNS_DELTA_FLAG))
+                || (!hook.hasPermission(AFTER_SWAP_FLAG) && hook.hasPermission(AFTER_SWAP_RETURNS_DELTA_FLAG))
+                || (!hook.hasPermission(AFTER_ADD_LIQ_FLAG) && hook.hasPermission(AFTER_ADD_LIQ_RETURNS_DELTA_FLAG))
+                || (!hook.hasPermission(AFTER_REMOVE_LIQ_FLAG) && hook.hasPermission(AFTER_REMOVE_LIQ_RETURNS_DELTA_FLAG))
+        ) return false;
         // If there is no hook contract set, then fee cannot be dynamic
         // If a hook contract is set, it must have at least 1 flag set, or have a dynamic fee
         return address(hook) == address(0)
             ? !fee.isDynamicFee()
-            : (uint160(address(hook)) >= AFTER_DONATE_FLAG || fee.isDynamicFee());
+            : (uint160(address(hook)) >= AFTER_REMOVE_LIQ_RETURNS_DELTA_FLAG || fee.isDynamicFee());
     }
 
-    /// @notice performs a hook call using the given calldata on the given hook
-    /// @return expectedSelector The selector that the hook is expected to return
-    /// @return selector The selector that the hook actually returned
-    function _callHook(IHooks self, bytes memory data) private returns (bytes4 expectedSelector, bytes4 selector) {
+    /// @notice performs a hook call using the given calldata on the given hook that doesnt return a delta
+    /// @return result The complete data returned by the hook
+    function callHook(IHooks self, bytes memory data) internal returns (bytes memory result) {
+        bool success;
+        (success, result) = address(self).call(data);
+        if (!success) _revert(result);
+
+        bytes4 expectedSelector;
+        bytes4 selector;
         assembly {
             expectedSelector := mload(add(data, 0x20))
+            selector := mload(add(result, 0x20))
         }
 
-        (bool success, bytes memory result) = address(self).call(data);
-        if (!success) _revert(result);
-
-        selector = abi.decode(result, (bytes4));
+        if (selector != expectedSelector) revert InvalidHookResponse();
     }
 
     /// @notice performs a hook call using the given calldata on the given hook
-    function callHook(IHooks self, bytes memory data) internal {
-        (bytes4 expectedSelector, bytes4 selector) = _callHook(self, data);
+    /// @return delta The delta returned by the hook
+    function callHookWithReturnDelta(IHooks self, bytes memory data, bool parseReturn)
+        internal
+        returns (int256 delta)
+    {
+        bytes memory result = callHook(self, data);
 
-        if (selector != expectedSelector) {
-            revert InvalidHookResponse();
-        }
+        if (!parseReturn) return 0;
+        (, delta) = abi.decode(result, (bytes4, int256));
```

Comment:
Shouldn't we still be requiring a valid selector be returned here?

Reply:
that happens in the call to `callHook` above